wordpress/wp-content/plugins/memberful-wp/src/acl.php

```diff
@@ -195,7 +214,116 @@ function memberful_wp_extract_slug_ids_and_user($args) {
+    return !empty( $user_subs );
+  }
+
+  // Get the set of restrictions for this post
+  $post_acl = get_post_meta( $post, 'memberful_acl', TRUE );
+  $plans_for_post = isset( $post_acl['subscription'] ) ? $post_acl['subscription'] : array();
+  $products_for_post = isset( $post_acl['product'] ) ? $post_acl['product'] : array();
+
+  // Get the term-level restrictions for all posts
+  $terms_acl = get_option( 'memberful_term_acl', array() );
+  $global_subscription_term_acl = isset( $terms_acl['subscription'] ) ? $terms_acl['subscription'] : array();
+  $global_product_term_acl = isset( $terms_acl['product'] ) ? $terms_acl['product'] : array();
+
+  // Find plans required to view the terms attached to this post
+  foreach ( $global_subscription_term_acl as $plan => $terms ) {
+    if ( array_intersect( $terms, $terms_for_post )) {
+      $plans_for_post[] = $plan;
+    }
+  }
+
+  // Find products required to view the terms attached to this post
+  foreach ( $global_product_term_acl as $product=> $terms ) {
+    if ( array_intersect( $terms, $terms_for_post )) {
+      $products_for_post[] = $product;
+    }
+  }
+
+  // Find any of the required plans that the current user has
+  $plan_intersect = array_intersect( $plans_for_post, $user_subs );
+
+  // Find any of the required products that the current user has
+  $user_products = $user ? array_keys( memberful_wp_user_products( $user )) : array();
+  $product_intersect = array_intersect( $products_for_post, $user_products );
+
+  if (( empty( $plans_for_post ) ) && ( empty( $products_for_post ))) {
+    // Grant access if no restrictions
+    return true;
+  } elseif ( ! empty( $plan_intersect ) || ! empty( $product_intersect )) {
+    // Grant access if any plans or products required and the user has at least one
+    return true;
+  } else {
+    // The post requires at least one plan or product to access and the user has none of those specified
+    return false;
+  }
+}
+
+function memberful_wp_post_viewable_by_any_registered_user( $post, $terms_for_post ) {
+  $posts_for_any_registered_users = memberful_wp_get_all_posts_available_to_any_registered_user();
+  $terms_for_any_registered_users = memberful_wp_get_all_terms_available_to_any_registered_user();
+
+  return ( in_array( $post, $posts_for_any_registered_users ) || array_intersect( $terms_for_post, $terms_for_any_registered_users ));
+}
+
+function memberful_wp_post_viewable_by_any_subscriber( $post, $terms_for_post ) {
+  $posts_for_anybody_subscribed_to_a_plan = memberful_wp_get_all_posts_available_to_anybody_subscribed_to_a_plan();
+  $terms_for_anybody_subscribed_to_a_plan = memberful_wp_get_all_terms_available_to_anybody_subscribed_to_a_plan();
+
+  return ( in_array( $post, $posts_for_anybody_subscribed_to_a_plan ) || array_intersect( $terms_for_post, $terms_for_anybody_subscribed_to_a_plan ));
+}
+
+function memberful_terms_restricting_post( $user, $post ) {
```

Comment:
@northerner If I'm correct, we use this function to find a term for displaying marketing content. We never need all the terms restricting a post, only the first one. Therefore we could maybe rename this function and return from it as soon as we find a term so we don't do unnecessary work.

This is not blocking at this point, but I'll try to create a PR with this change.

Reply:
https://github.com/memberful/memberful-wp/pull/299